### ISSUE TITLE
read local xml files

### ISSUE THREAD

**Issue by Phu2** [ropensci/oai#59]
It would be nice if you could read the local xml files that were downloaded by using `dump_raw_to_txt` or another harvester. I imagine a function `list_local_records(startfile = "/path/to/file", prefix = "oai_dc")`, which reads all the local dump files using the resumptionToken in the startfile. Similar to `list_records` but for local files.

**Comment by sckott** [comment]
Thanks. I'll have a look

**Comment by sckott** [comment]
@Phu2 I'm not sure this falls within scope for this pkg. can you explain your use case a bit more and pain points encountered? what's hard about the current setup?  is it hard to find the local files? hard to parse them? something else?

**Comment by Phu2** [comment]
Sorry for the delay. I was happy about the possibility to download the records for later use. Therefore the import of these local files seems to be a logical step to me. Sometimes i get these files from colleagues or via other tools. It's this kind of offline mode of work i am looking for.

Parsing the files is another concern. oai_dc is the mandatory metadata format but also sparse and imprecise at the same time. I would always prefer other formats like Mods or Datacite. Being able to read those formats would be a big plus in terms of convenience.

For now i'm using XSLT for the preprocessing, and that's perfectly fine.

**Comment by sckott** [comment]
Okay, thanks. 

> which reads all the local dump files using the resumptionToken in the startfile

What is the startfile? And why would you use a resumptionToken locally?

**Comment by mbojan** [comment]
Is the "startfile" the first XML of the "multipage" result set?

I believe the package does next to nothing in terms of parsing the results, and I think it should stay this way. The "parsing" of XMLs which is performed if `as = "df"` is essentially, AFAIR, flattening the result with `tibble::as_tibble()`. This means that you could, RAM-permitting, try something along the following almost-pseudo-code:

```r
list_of_xml_file_names_from_a_oai_request %>%
  lapply(xml2::read_xml) %>%
  lapply(tibble::as_tibble) %>%
  dplyr::bind_rows()
```

... unless I am missing something...

An alternative, which probably makes sense if you are in for more or repetitive data fetching, is to write a custom dumper which will parse the XML on the fly extracting the information, assembling it in a form you need, writing to disk/database etc...

**Comment by Phu2** [comment]
> What is the startfile?
> Is the "startfile" the first XML of the "multipage" result set?

Yes, it is. If you get multiple XML files from a `list_records` command it is the one without a resumptionToken. All subsequent XML files have a resumptionToken in the `<request>`-Element. 

> And why would you use a resumptionToken locally?

Well, i was thinking of keeping the order of the responses aka the [list response sequence](http://www.openarchives.org/OAI/openarchivesprotocol.html#FlowControl). But this is probably negligible as each record must have an identifier and be complete.

> I believe the package does next to nothing in terms of parsing the results

if you are working with the `oai_dc` format, the current approach is fine because of the flat structure within the `<metadata>` element. For instance:

```
...
<record>
      <header>
        ....
      </header>
      <metadata>
        <oai_dc:dc xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/" xmlns:doc="http://www.lyncode.com/xoai" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dc="http://purl.org/dc/elements/1.1/" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
          <dc:title>Commercialization of academic research</dc:title>
          <dc:creator>Houweling, Stefan</dc:creator>
          <dc:subject>University spin-off</dc:subject>
          <dc:subject>Commercialization</dc:subject>
          <dc:subject>Academic research</dc:subject>
          <dc:subject>Biotech</dc:subject>
          <dc:subject>ddc:330</dc:subject>
          <dc:description>The central research ....</dc:description>
          <dc:date>2017-01-01</dc:date>
          <dc:type>doctoralThesis</dc:type>
          <dc:type>Text</dc:type>
          <dc:type>publishedVersion</dc:type>
          <dc:identifier>https://dspace.ub.uni-siegen.de/handle/ubsi/1216</dc:identifier>
          <dc:identifier xsi:type="urn:nbn">urn:nbn:de:hbz:467-12164</dc:identifier>
          <dc:language>eng</dc:language>
          <dc:rights>info:eu-repo/semantics/openAccess</dc:rights>
          <dc:rights>https://dspace.ub.uni-siegen.de/static/license.txt</dc:rights>
          <dc:publisher>Universität Siegen</dc:publisher>
          <dc:publisher>Fakultät III Wirtschaftswissenschaften, Wirtschaftsinformatik und Wirtschaftsrecht</dc:publisher>
        </oai_dc:dc>
      </metadata>
    </record>
...
```

Here is a record in the `mets` format with a much deeper nesting of metadata elements.

```
...
<record>
      <header>
        ...
      </header>
      <metadata>
        <mets xmlns="http://www.loc.gov/METS/" xmlns:doc="http://www.lyncode.com/xoai" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.loc.gov/METS/ http://www.loc.gov/standards/mets/mets.xsd" PROFILE="DSpace METS SIP Profile 1.0" TYPE="DSpace ITEM" ID="&#10;&#9;&#9;&#9;&#9;DSpace_ITEM_ubsi-1218" OBJID="&#10;&#9;&#9;&#9;&#9;hdl:ubsi/1218">
          <metsHdr CREATEDATE="2020-09-13T23:04:20Z">
            <agent TYPE="ORGANIZATION" ROLE="CUSTODIAN">
              <name>OPUS Siegen</name>
            </agent>
          </metsHdr>
          <dmdSec ID="DMD_ubsi_1218">
            <mdWrap MDTYPE="MODS">
              <xmlData xmlns:mods="http://www.loc.gov/mods/v3" xsi:schemaLocation="http://www.loc.gov/mods/v3 http://www.loc.gov/standards/mods/v3/mods-3-1.xsd">
                <mods:mods xsi:schemaLocation="http://www.loc.gov/mods/v3 http://www.loc.gov/standards/mods/v3/mods-3-1.xsd">
                  <mods:name>
                    <mods:role>
                      <mods:roleTerm type="text">author</mods:roleTerm>
                    </mods:role>
                    <mods:namePart>Baumgart, Ingo</mods:namePart>
                  </mods:name>
                  <mods:extension>
                    <mods:dateAccessioned encoding="iso8601">2019-09-02T10:04:31Z</mods:dateAccessioned>
                  </mods:extension>
                  <mods:extension>
                    <mods:dateAvailable encoding="iso8601">2017-10-17T12:12:12Z</mods:dateAvailable>
                  </mods:extension>
                  <mods:originInfo>
                    <mods:dateIssued encoding="iso8601">2017</mods:dateIssued>
                  </mods:originInfo>
                  <mods:identifier type="uri">https://dspace.ub.uni-siegen.de/handle/ubsi/1218</mods:identifier>
                  <mods:identifier type="urn">urn:nbn:de:hbz:467-12182</mods:identifier>
                  <mods:abstract>...</mods:abstract>
                  <mods:language>
                    <mods:languageTerm authority="rfc3066">de</mods:languageTerm>
                  </mods:language>
                  <mods:accessCondition type="useAndReproduction"/>
                  <mods:titleInfo>
                    <mods:title>Erzeugung von robusten magnetfeldabhängigen Qubitzuständen für hochpräzise Magnetometrie</mods:title>
                  </mods:titleInfo>
                  <mods:genre>Doctoral Thesis</mods:genre>
                </mods:mods>
              </xmlData>
            </mdWrap>
          </dmdSec>
          <amdSec ID="TMD_ubsi_1218">
            <rightsMD ID="RIG_ubsi_1218">
              <mdWrap OTHERMDTYPE="DSpaceDepositLicense" MDTYPE="OTHER" MIMETYPE="text/plain">
                <binData>...</binData>
              </mdWrap>
            </rightsMD>
          </amdSec>
          <amdSec ID="FO_ubsi_1218_1">
            <techMD ID="TECH_O_ubsi_1218_1">
              <mdWrap MDTYPE="PREMIS">
                <xmlData xmlns:premis="http://www.loc.gov/standards/premis" xsi:schemaLocation="http://www.loc.gov/standards/premis http://www.loc.gov/standards/premis/PREMIS-v1-0.xsd">
                  <premis:premis>
                    <premis:object>
                      <premis:objectIdentifier>
                        <premis:objectIdentifierType>URL</premis:objectIdentifierType>
                        <premis:objectIdentifierValue>https://dspace.ub.uni-siegen.de/bitstream/ubsi/1218/1/Dissertation_Ingo_Baumgart.pdf</premis:objectIdentifierValue>
                      </premis:objectIdentifier>
                      <premis:objectCategory>File</premis:objectCategory>
                      <premis:objectCharacteristics>
                        <premis:fixity>
                          <premis:messageDigestAlgorithm>MD5</premis:messageDigestAlgorithm>
                          <premis:messageDigest>383566bc3e8b5a4fc38c933ddf96d720</premis:messageDigest>
                        </premis:fixity>
                        <premis:size>167739931</premis:size>
                        <premis:format>
                          <premis:formatDesignation>
                            <premis:formatName>application/pdf</premis:formatName>
                          </premis:formatDesignation>
                        </premis:format>
                      </premis:objectCharacteristics>
                      <premis:originalName>Dissertation_Ingo_Baumgart.pdf</premis:originalName>
                    </premis:object>
                  </premis:premis>
                </xmlData>
              </mdWrap>
            </techMD>
          </amdSec>
          <fileSec>
            <fileGrp USE="ORIGINAL">
              <file ID="BITSTREAM_ORIGINAL_ubsi_1218_1" MIMETYPE="application/pdf" SEQ="1" SIZE="167739931" CHECKSUM="383566bc3e8b5a4fc38c933ddf96d720" CHECKSUMTYPE="MD5" ADMID="FO_ubsi_1218_1" GROUPID="GROUP_BITSTREAM_ubsi_1218_1">
                <FLocat xlink:type="simple" LOCTYPE="URL" xlink:href="https://dspace.ub.uni-siegen.de/bitstream/ubsi/1218/1/Dissertation_Ingo_Baumgart.pdf"/>
              </file>
            </fileGrp>
          </fileSec>
          <structMap TYPE="LOGICAL" LABEL="DSpace Object">
            <div TYPE="DSpace Object Contents" ADMID="DMD_ubsi_1218">
              <div TYPE="DSpace BITSTREAM">
                <fptr FILEID="BITSTREAM_ORIGINAL_ubsi_1218_1"/>
              </div>
            </div>
          </structMap>
        </mets>
      </metadata>
    </record>
...
```

So, if you do something like

```
test <- oai::list_records(oai_url, prefix = 'mets')
```  

all the descriptive metadata wihtin the `dmdSec` element will be munged into one variable.

**Comment by sckott** [comment]
@Phu2 Can you give an example of a OAI instance with that `mets` format example above?

**Comment by Phu2** [comment]
The example above stems from https://dspace.ub.uni-siegen.de/oai/request. More OAI endpoints supporting mets:

- http://oai.openedition.org
- https://services.ub.uni-leipzig.de/digitalcollections/oai2

Other elaborated formats:

- https://api.europeana.eu/oai/record?verb=GetRecord&identifier=http%3A%2F%2Fdata.europeana.eu%2Fitem%2F114%2Fdata_sounds_SA_69_1_12_1&metadataPrefix=edm
- https://www.biodiversitylibrary.org/oai?verb=ListRecords&metadataPrefix=olef
- http://www.mimo-db.eu:8080/oaicat/OAIHandler?verb=ListRecords&metadataPrefix=lido&set=MU

**Comment by mbojan** [comment]
@sckott I believe there is very little in terms of standardization of how the actual records look like. Perhaps beyond Dublin Core (DC), but I may be wrong. My initial sentiment is that it is a failed quest to try to support them. Especially that I can imagine developers might introduce countless variations (we actually did a variant of DC for now defunct Polish Scholarly Bibliography).

**Comment by sckott** [comment]
I agree if there is too much variation that we shouldn't try to support this. @Phu2 do you have a sense for how common mets is among data providers with OAI?

one concern i have is with deeply nested data, how do we know how to parse it? how do we know what data users want?  Seems the only approach we could take here is to parse everything.

**Comment by mbojan** [comment]
> one concern i have is with deeply nested data, how do we know how to parse it? how do we know what data users want? Seems the only approach we could take here is to parse everything.

Exactly.

If there is a need for "custom" parsing of the XMLs from a particular OAI provider (beyond generic flattening of everything to a `tibble`), perhaps it will make sense to have it as a separate package importing this one, won't it?

**Comment by sckott** [comment]
Yeah, or custom scripts by each person.

**Comment by mbojan** [comment]
Not really in the scope of the package.
